core/src/main/scala/neotypes/model.scala

```diff
@@ -0,0 +1,198 @@
+package neotypes
+package model
+
+import mappers.{ParameterMapper, ResultMapper}
+
+import org.neo4j.driver.types.{IsoDuration => NeoDuration, Point => NeoPoint}
+
+import java.time.{LocalDate => JDate, LocalDateTime => JDateTime, LocalTime => JTime, OffsetTime => JZTime, ZonedDateTime => JZDateTime}
+import java.util.{Map => JMap}
+import scala.collection.immutable.ArraySeq
+import scala.jdk.CollectionConverters._
+import scala.util.control.NoStackTrace
+
+/** Safe wrapper over a Neo4j query parameter. */
+final class QueryParam private[neotypes] (private[neotypes] val underlying: AnyRef) extends AnyVal
+object QueryParam {
+  def apply[A](scalaValue: A)(implicit mapper: ParameterMapper[A]): QueryParam =
+    mapper.toQueryParam(scalaValue)
+
+  private[neotypes] def toJavaMap(map: Map[String, QueryParam]): JMap[String, Object] =
+    map.map {
+      case (key, value) =>
+        key -> value.underlying
+    }.asJava
+}
+
+/** Data types supported by Neo4j. */
+object types {
+  /** Parent type of all Neo4j types. */
+  sealed trait NeoType extends Product with Serializable
+
+  /** Parent type of all Neo4j types that have named properties. */
+  sealed trait NeoObject extends NeoType {
+    def properties: Map[String, NeoType]
+
+    final def get(key: String): Option[NeoType] =
+      properties.get(key)
+
+    final def getAs[T](key: String)(implicit mapper: ResultMapper[T]): Either[exceptions.ResultMapperException, T] =
+      properties
+        .get(key)
+        .toRight(left = exceptions.PropertyNotFoundException(key))
+        .flatMap(mapper.decode)
+
+    final def keys: Set[String] =
+      properties.keySet
+
+    final def values: Iterable[NeoType] =
+      properties.values
+  }
+
+  /** Represents a Neo4j heterogeneous list (composite type) */
+  final case class NeoList(values: Iterable[NeoType]) extends NeoType
+
+  /** Represents a Neo4j heterogeneous map (composite type) */
+  final case class NeoMap(properties: Map[String, NeoType]) extends NeoObject
+
+  /** Parent type of all Neo4j structural types. */
+  sealed trait Entity extends NeoObject {
+    def elementId: String
+
+    override def properties: Map[String, Value]
+  }
+
+  /** Represents a Neo4j Node. */
+  final case class Node(
+      elementId: String,
+      labels: Set[String],
+      properties: Map[String, Value]
+  ) extends Entity {
+    /** Checks if this Node contains the given label; case insensitive. */
+    def hasLabel(label: String): Boolean =
+      labels.contains(label.trim.toLowerCase)
```

Comment:
Should labels be a case-insensitive string?

https://github.com/typelevel/case-insensitive

Reply:
Yeah, sadly **neotypes** has a strong policy of only depending on the **neo4j-java-driver** for its `core` module.
Even **Shapeless** was moved to the `generic` module.